### ISSUE TITLE
`parallel` tasks should have the option to kill children on exit

### ISSUE THREAD

**Issue by ezracelli** [sagiegurari/cargo-make#751]
### Feature Description

Currently, when a child of a a `parallel` task exits &hellip;

| &hellip; with an exit code of zero | &hellip; with a non-zero exit code |
| --- | --- |
| the runner continues as if nothing happened (without any logging, unless `fork` is enabled) | the runner exits (with an error message + exit code of `1`) &mdash; without killing any still-living child tasks |

<details>
<summary><code>Makefile.toml</code></summary>

```toml
[tasks."parent"]
[tasks."parent".run_task]
# fork = true
name = ["child:1", "child:2"]
parallel = true

[tasks."child:1"]
script = '''
#!/usr/bin/env bash

sleep 1
# toggle below to trigger a non-zero exit code
# false
'''

[tasks."child:2"]
script = '''
#!/usr/bin/env bash

sleep 2
'''
```
</details>

This is mostly an issue for long-lived tasks, like filesystem watchers. My use case is two long-lived tasks; if one fails, the other continues in the background, and I need to remember to `ps aux | grep` and `kill` manually.

---

I've used [`concurrently`](https://www.npmjs.com/package/concurrently) extensively, and I really like its implementation. It provides `--kill-others` and `--kill-others-on-fail` flags (and a corresponding `killOthers: ("failure" | "success")[]` API option).

* Without `--kill-others[-on-fail]`, when a child exits &hellip;

  | &hellip; with an exit code of zero | &hellip; with a non-zero exit code |
  | --- | --- |
  | the runner continues as if nothing happened (without any logging) &mdash; _this is the same behavior as `cargo-make`_ | the runner continues as if nothing happened (without any logging), then exits with a code of `1` |
* With `--kill-others`, when a child exits &hellip;

  | &hellip; with an exit code of zero | &hellip; with a non-zero exit code |
  | --- | --- |
  | the runner exits with a code of `1`, killing other children with `SIGTERM` | the runner exits with a code of `1`, killing other children with `SIGTERM` |
* With `--kill-others-on-fail`, when a child exits &hellip;

  | &hellip; with an exit code of zero | &hellip; with a non-zero exit code |
  | --- | --- |
  | the runner continues as if nothing happened (without any logging) &mdash; _this is the same behavior as `cargo-make`_ | the runner exits with a code of `1`, killing other children with `SIGTERM` |

### Describe The Solution You'd Like

I think the current implementation is mostly fine, with the exception of not killing children on exit. In my opinion, there's two potential simpler solutions that wouldn't involve adding configuration options:

When the child of a `parallel` task exits with a non-zero exit code, &hellip;

* &hellip; do nothing; then exit with the existing error message and exit code of `1` when all other children exit
* &hellip; exit immediately and kill all living children

This wouldn't solve _my_ use-case, but it at least would resolve the existing bug.

---

Higher-effort would be to implement an option similar to `concurrently`, maybe nesting it under the existing `parallel` option:

```toml
parallel = { kill_others = ["failure", "success"] }
```

or, if we want to be able to specify the signal sent to living children (though this might be difficult to implement cross-platform), maybe:

```toml
parallel = { on_error = "SIGKILL", on_exit = "SIGTERM" }
```

or even specifying a task (that would be provided with the info of the process to be killed) to enable arbitrary actions:

```toml
[task."name"]
parallel = { on_error_task = "name:on_error" }

[task."name:on_error"]
script = '''
#!/usr/bin/env bash

notify-send -u critical "killing task ${CARGO_MAKE_TASK_NAME}"
kill "${CARGO_MAKE_TASK_PID}"
'''
```

---

Happy to draft up a PR, just curious which direction (if any) the maintainers would like to see this feature go!

**Comment by sagiegurari** [comment]
@ezracelli thanks a lot for the detailed info.
question, did you look at the cleanup task capability? maybe we can use that to 'kill' the background process?

from this section:
https://github.com/sagiegurari/cargo-make#commands-scripts-and-sub-tasks

```toml
[tasks.cleanup]
command = "echo"
args = ["cleanup"]

[tasks.cleanup-example]
run_task = { name = ["echo1", "echo2", "fail"], fork = true, cleanup_task = "cleanup" }
```

**Comment by sagiegurari** [comment]
@ezracelli did it solve your issue?

**Comment by ezracelli** [comment]
@sagiegurari Thanks for the tip! I did a bit of experimentation with `cleanup_task`, but unfortunately as it doesn't know what the PIDs of the children are, it requires maintaining a project-local `/run`-type directory with pidfiles (so the `cleanup` task can kill them). This isn't a cross-platform solution, though, as Windows doesn't do PIDs the same way *NIX does. My solution, on *NIX:

```sh
mkdir -p ./.run/
echo $$ > ./.run/process.pid
exec process
```

&hellip; and the `cleanup_task`:

```sh
if [ -d ./.run ]; then
  cat ./.run/*.pid | xargs kill
  rm ./.run/*.pid
fi
```

Obviously, this isn't portable to e.g. Windows, so this solution would require a separate Windows-specific `makefile.toml` if a project wanted to support cross-platform task running.

In my opinion, leaving zombie child processes running (the current behavior) is actually a bug. `cargo-make` is amazing, this is the one thing that feels like it's not fully fleshed out :grin:

**Comment by sagiegurari** [comment]
i'm trying to see if we have a workaround that is quick without having to do pretty complex stuff inside cargo-make.
for example, you said you would need different makefile for windows, but actually you have a way to extend specific tasks per platform: https://github.com/sagiegurari/cargo-make#usage-platform-override 

anyhow, cargo-make comes with duckscript embedded.
so you could use 'spawn' to call an executable, get its pid and store that in an env var (using set_env command).
than in cleanup task, call exec with kill by pid you stored in the env var.

if this is a pattern that happens alot, you could setup a plugin that does all that for you.
https://github.com/sagiegurari/cargo-make#usage-plugins
plugins are a way to give cargo-make more power without coding it inside cargo-make.

i agree that leaving zombie child processes isn't great. that can happen for many different tools and usually child processes are killed not by parent process but by the OS unless they probably prevent that or something.
while solving your specific usecase might be 'low-med' effort, there are other more complex scenarios that could lead to real complexity. for example while if child started a child on its own but than is killed by cargo-make (parent) process? who kills the sub child?
so i prefer to see if we can get to a solution for your specific use case without building a feature that might break for others.

**Comment by scottbot95** [comment]
> did it solve your issue?

I would argue that *even if* the cleanup task suggestion works, it would be a workaround at best and not a solution. I agree with @ezracelli that it very much *feels* like a bug, however I can see how this would be "working as designed". IMO at a minimum there should be a not in the (tiny) section about parallel tasks clarifying this fairly un-intuitive behavior.

**Comment by sagiegurari** [comment]
@scottbot95 thanks for the feedback.

> IMO at a minimum there should be a not in the (tiny) section about parallel tasks clarifying this fairly un-intuitive behavior.

would you like to submit a PR that explains it better? if you do, please make sure to update the content.md which the README is generated from

> Also, unless I'm missing something, you cannot provide a cleanup task with a duckscript task so your proposed duckscript solution would require at a minimum 3 tasks (cleanup, the duckscript task, and a wrapper parent with run_task = { name = [ "duckscript"], cleanup_task = "cleanup" })

i didn't get this part. you already have 2 tasks, right?
the task that starts up at least one task in parallel and the task that is invoked from it
now we need to
1. add a cleanup task
2. modify the invoked task to not be command+args but be duckscript that invokes the command and args so you can keep the pid in some env var. so its not adding another task, its replace how you would right the original task

ya i get its not amazing, but its a tricky area that can lead to many edge cases.

**Comment by scottbot95** [comment]
> would you like to submit a PR that explains it better?

I am travelling currently, but I can work on an update when I get back home in a day or two. I don't think there needs to be much.

---

> i didn't get this part ....

Ahhhh yeah that makes sense I was just confused. Thanks!

**Comment by sagiegurari** [comment]
documentation updated
